feat: Allow traefik merged static configuration override with `static_config.*.yaml`

rfay opened the pull request:
## The Issue

This is a followup to 
* https://github.com/ddev/ddev/pull/6317

We discovered there that we didn't have an appropriate way to add or override static configuration for traefik. We could only add to the base config, couldn't merge or override any elements. Now we can merge and override. For example, several plugins can be added in separate files, logLevel can be changed, etc.

## How This PR Solves The Issue

`~/.ddev/traefik/static_config.*.yaml` can now override and is generated into `~/.ddev/traefik/.static_config.yaml`

This will make it possible to do things like changing log, overriding plugins, etc.

## Manual Testing Instructions

Docs review:
* https://ddev--6379.org.readthedocs.build/en/6379/users/extend/traefik-router/#traefik-configuration
* https://ddev--6379.org.readthedocs.build/en/6379/users/topics/hosting/

Try the examples for static_config.*.yaml in the [traefik router docs](https://ddev--6379.org.readthedocs.build/en/6379/users/extend/traefik-router)

## Automated Testing Overview

* TestTraefikStaticConfig
* TestMergeYaml

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

rfay commented on docs/content/users/extend/traefik-router.md:
```diff
@@ -21,10 +21,32 @@ All Traefik configuration uses the *file* provider, not the *Docker* provider. E
+
 * `certs/default_cert.*` files are the default DDEV-generated certificates, normally created by `mkcert`.
-* `config/default_config.yaml` contains global dynamic configuration, including pointers to the default certificates. It is possible to add other Traefik configuration in the `config` directory, which will apply to all projects. For example, a `config/router_middlewares.yaml` file might provide middleware implementations that would apply to all projects.
+* `config/default_config.yaml` contains global *dynamic* configuration, including pointers to the default certificates. It is possible to add other Traefik configuration in the `config` directory, which will apply to all projects. For example, a `config/router_middlewares.yaml` file might provide middleware implementations that would apply to all projects.
```
> Do these get merged in the same fashion as the static_config.*.yaml files? If so, should they follow the same xyz.*.yaml format?

No, no merging, they're part of the dynamic configuration, and nothing done with them. I didn't really realize how/why they might be useful until talking to you one day.

> Finally, if they are defined here, can they be overridden by the project-specific dynamic config? I assume so, but it might be worth stating it.

I think it's probably alphabetic. You can easily see what goes on by `docker exec -it ddev-router bash` and `cd config`. You'll see these default_config files there. You can edit them. My bet is that `ls -l` in that directory is meaningful.  But I also bet they only do things nobody else does.

I'm not completely sure why we ended up with these default_config, this isn't specifically a Traefik idea, seems to be a DDEV idea. Normally only has the default certs in it, and normally the default certs aren't used.

rfay commented on docs/content/users/extend/traefik-router.md:
```diff
@@ -21,10 +21,32 @@ All Traefik configuration uses the *file* provider, not the *Docker* provider. E
-* `.static_config.yaml` is the base configuration, generated by DDEV, and not to be edited. It is read on router startup, and does not change after that.
-* Additional static configuration may be added by adding a `static_config.*.yaml` file, which will be appended to the generated `.static_config.yaml`.
+* `.static_config.yaml` (a hidden file) is the configuration that gets used. It is not to be edited; it is generated from DDEV's base configuration while merging any files named `static_config.*.yaml`. It is read on router startup, and does not change until the router starts again (normally after `ddev poweroff`).
+* Additional static configuration may be added by adding `static_config.*.yaml` files, which will be merged into the generated `.static_config.yaml`. For example, a `static_config.plugin.yaml` might contain external Traefik plugins, or a `static_config.dnschallenge.yaml` might provide configuration for additional `certificatesResolvers`. Merging is done with an **override** strategy, meaning that ...
```
```suggestion
* Additional static configuration may be added by adding `static_config.*.yaml` files, which will be merged into the generated `.static_config.yaml`. For example, a `static_config.plugin.yaml` might contain external Traefik plugins, or a `static_config.dnschallenge.yaml` might provide configuration for additional `certificatesResolvers`. Merging is done with an **override** strategy, meaning that the final file in alphanumeric sort to touch a particular element of the YAML structure wins.
```

Just minor adaptation of your version.

rfay commented:
Thanks for the review!

rfay commented:
@nickchomey - the loss of comments is not a problem with mergo (or at least doesn't start there). It's a problem with go-yaml. I've never figured out how to get past it, might be worth a return to it.

The easiest way to see it is to create a comment in your config.yaml and `ddev config --auto`, comment will be lost.

See
* https://github.com/go-yaml/yaml/issues/132#issuecomment-498791474
* https://stackoverflow.com/questions/56401238/how-to-unmarshall-yaml-v3-with-full-access-to-node
* Many problems in https://github.com/go-yaml/yaml/issues?q=is%3Aopen+is%3Aissue+in%3Atitle+comment+
* https://github.com/ddev/ddev/issues/1529
* https://github.com/ddev/ddev/issues/1074

stasadev reviewed (approved):
I tested, and it works great!
This is the exact thing that I was expecting.

@rfay could you change [`sendAnonymousUsage`](https://doc.traefik.io/traefik/contributing/data-collection/) to camelCase in a few places to match the official naming:

```diff
-sendanonymoususage: false
+sendAnonymousUsage: false
```

rfay commented:
Wow, thanks for catching that sendAnonymousUsage @stasadev . I guess it must have been that way for years. 

Rebased, fixed that, added variation of @nickchomey suggestion. When most tests pass I'll pull and @nickchomey can rebase against HEAD.

seebeen commented:
I'm buying you a beer for this when I'm in the US!
